python/lsst/cp/pipe/deferredCharge.py

```diff
@@ -0,0 +1,1082 @@
+        calib : `lsst.ip.isr.DeferredChargeCalib`
+            Calibration to populate with values.
+        detector : `lsst.afw.cameraGeom.Detector`
+            Detector object containing the geometry information for
+            the amplifiers.
+
+        Returns
+        -------
+        calib : `lsst.ip.isr.DeferredChargeCalib`
+            Populated calibration.
+
+        Notes
+        -----
+        The original CTISIM code uses a data model in which the
+        "overscan" consists of the standard serial overscan bbox with
+        the values for the last imaging data column prepended to that
+        list.  This version of the code keeps the overscan and imaging
+        sections separate, and so a -1 offset is needed to ensure that
+        the same columns are used for fitting between this code and
+        CTISIM.  This offset removes that last imaging data column
+        from the count.
+        """
+        # Range to fit.  These are in "camera" coordinates, and so
+        # need to have the count for last image column removed.
+        start, stop = self.config.globalCtiColumnRange
+        start -= 1
+        stop -= 1
+
+        # Loop over amps/inputs, fitting those columns from
+        # "non-saturated" inputs.
+        for amp in detector.getAmplifiers():
+            ampName = amp.getName()
+
+            # Number of serial shifts.
+            nCols = amp.getRawDataBBox().getWidth() + amp.getRawSerialPrescanBBox().getWidth()
+
+            # The signal is the mean intensity of each input, and the
+            # data are the overscan columns to fit.  For detectors
+            # with non-zero CTI, the charge from the imaging region
+            # leaks into the overscan region.
+            signal = []
+            data = []
+            Nskipped = 0
+            for exposureEntry in inputMeasurements:
+                exposureDict = exposureEntry['CTI']
+                if exposureDict[ampName]['IMAGE_MEAN'] < self.config.maxSignalForCti:
+                    signal.append(exposureDict[ampName]['IMAGE_MEAN'])
+                    data.append(exposureDict[ampName]['OVERSCAN_VALUES'][start:stop+1])
+                else:
+                    Nskipped += 1
+            self.log.info(f"Skipped {Nskipped} exposures brighter than {self.config.maxSignalForCti}.")
+
+            signal = np.array(signal)
+            data = np.array(data)
+
+            ind = signal.argsort()
+            signal = signal[ind]
+            data = data[ind]
+
+            # CTI test.  This looks at the charge that has leaked into
+            # the first few columns of the overscan.
+            overscan1 = data[:, 0]
+            overscan2 = data[:, 1]
+            test = (np.array(overscan1) + np.array(overscan2))/(nCols*np.array(signal))
+            testResult = np.median(test) > 5.E-6
+            self.log.info("Estimate of CTI test is %f for amp %s, %s.", np.median(test), ampName,
```

Comment:
I see three `%` on this line and only two variables after the string closes. This is why I hate this kind of formatting and think we've got it all wrong, as this sort of mistake can't happen with f-strings. Am I misreading this due to my inexperience with this type of formatting, or is that an error?

Reply:
Oh god, it's the following string that's being added in with that last arg 😳

Reply:
Sorry for the noise, but this is why I hate this way of doing things.

Reply:
TBH, what I really wanted to write was `testResult ? str1 : str2`, but that's not valid python.